### PR TITLE
[FLINK-19700][k8s] Make Kubernetes Client in KubernetesResourceManagerDriver use io executor

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently, the Kubernetes Client in `KubernetesResourceManagerDriver` is using a dedicated thread pool through calling `KubeClientFactory::createThreadPoolForAsyncIO` to create fixed thread pool. `KubernetesResourceManagerDriver` could get the io executor and eliminate the redundant thread pool. `KubeClientFactory` uses `KubeClientFactory::createThreadPoolForAsyncIO` by default.*

## Brief change log

  - *`KubernetesResourceManagerDriver` adds constructor with creating `FlinkKubeClient ` through `KubeClientFactory#fromConfiguration(flinkConfig, ioExecutor)` to reuse the io executor of `AbstractResourceManagerDriver`.*

## Verifying this change

  - *`TestingResourceManagerDriver` updates `initialize()` with `ExecutorService` type parameter to verify the `AbstractResourceManagerDriver` whether to work.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by SteNicholas** [comment]
@xintongsong @wangyang0918, could you please help to review the pull request if you are available?

File: flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DefaultDispatcherResourceManagerComponentFactory.java [hunk 1]
```diff
@@ -69,7 +69,7 @@
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
```
Comment (xintongsong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DefaultDispatcherResourceManagerComponentFactory.java:72]: I'm not sure about replacing `Executor` with `ExecutorService` for all these places.

Looking into `Fabric8FlinkKubeClient`, I think the only reason `FlinkKubeClient` uses `ExecutorService` rather than `Executor` is that, it will shutdown and release the thread pool in `close`. That is because the client uses a dedicated thread pool, and if it does not shutdown the thread pool, other components won't do that either.

Since now the client uses a shared IO executor, it should no longer shutdown the executor on close. Otherwise, all the other components using that IO executor would be affected. That means we can replace `ExecutorService` with `Executor` for `FlinkKubeClient`.

**Review by xintongsong — CHANGES_REQUESTED** [review]
Thanks for preparing this PR, @SteNicholas.
I'm afraid the changes won't work. Please see my comments. The e2e failures on AZP also confirmed my concern.

**Comment by SteNicholas** [comment]
> Thanks for preparing this PR, @SteNicholas.
> I'm afraid the changes won't work. Please see my comments. The e2e failures on AZP also confirmed my concern.

@xintongsong , thanks for your detailed comments. I have followed your comments mentioned above. Please help to review again if you are available.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java [hunk 2]
```diff
@@ -64,7 +65,7 @@
 
 	private final Time podCreationRetryInterval;
 
-	private final FlinkKubeClient kubeClient;
+	private FlinkKubeClient kubeClient;
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:68]:
```suggestion
	private Optional<FlinkKubeClient> kubeClientOpt;
```
The client is used a lot in this driver. I would suggest to use `Optional` here to force a existence check whenever it is used.

Comment (xintongsong) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:68]: Actually, we should also have such protection for `podsWatch`. But it is not a problem of this PR.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java [hunk 3]
```diff
@@ -85,23 +86,31 @@
...
 		this.podCreationCoolDown = FutureUtils.completedVoidFuture();
 	}
 
+	public KubernetesResourceManagerDriver(
+			Configuration flinkConfig,
+			FlinkKubeClient kubeClient,
+			KubernetesResourceManagerDriverConfiguration configuration) {
+		this(flinkConfig, configuration);
+		this.kubeClient = Preconditions.checkNotNull(kubeClient);
+	}
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:103]: Instead of introducing another constructor, I think we can have only one constructor that takes `KubeClientFactory` as an argument. We would also need to make `KubeClientFactory` and interface and provide a singleton implementation.
In this way, the tests can provide a testing factory implementation, without needing to have a separate code path in the production codes.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java [hunk 4]
```diff
@@ -85,23 +86,31 @@
...
+
 	// ------------------------------------------------------------------------
 	//  ResourceManagerDriver
 	// ------------------------------------------------------------------------
 
 	@Override
 	protected void initializeInternal() throws Exception {
+		if (kubeClient == null) {
+			kubeClient = KubeClientFactory.fromConfiguration(flinkConfig, getIoExecutor());
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:112]: Currently this line of code cannot be covered by the unit tests, due to usage of a different constructor.

**Review by xintongsong — COMMENTED** [review]
Thanks @SteNicholas. I think the changes should work now.
I have a few more comments concerning the maintainability and testability. Please take a look.

**Comment by SteNicholas** [comment]
> Thanks @SteNicholas. I think the changes should work now.
> I have a few more comments concerning the maintainability and testability. Please take a look.

@xintongsong Please help to review the pull request again. Sorry for trouble you again.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java [hunk 5]
```diff
@@ -41,6 +42,8 @@
 
 	private static final String CLUSTER_ID_PREFIX = "flink-cluster-";
 
+	private static final KubeClientFactory kubeClientFactory = DefaultKubeClientFactory.getInstance();
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterClientFactory.java:45]: minor: this can be a local variable

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java [hunk 6]
```diff
@@ -75,7 +80,7 @@
 	/** Current max pod index. When creating a new pod, it should increase one. */
 	private long currentMaxPodId = 0;
 
-	private KubernetesWatch podsWatch;
+	private Optional<KubernetesWatch> podsWatch;
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:83]: The `podsWatch` related changes should be in a separate commit. They are hotfix for exist codes and is unrelated to the purpose of this PR.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java [hunk 7]
```diff
@@ -158,7 +165,9 @@ public void deregisterApplication(ApplicationStatus finalStatus, @Nullable Strin
 		// In case of pod creation failures, we should wait for an interval before trying to create new pods.
 		// Otherwise, ActiveResourceManager will always re-requesting the worker, which keeps the main thread busy.
 		final CompletableFuture<Void> createPodFuture =
-				podCreationCoolDown.thenCompose((ignore) -> kubeClient.createTaskManagerPod(taskManagerPod));
+				podCreationCoolDown.thenCompose((ignore) -> kubeClient
+						.map(client -> client.createTaskManagerPod(taskManagerPod))
+						.orElse(CompletableFuture.completedFuture(null)));
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java:170]: Since we are expecting the client to be non-empty, I think we should fail explicitly for the empty case. 
Same for the other occurrences.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java [hunk 8]
```diff
@@ -64,6 +65,8 @@
 		"stop - stop the kubernetes cluster\n" +
 		"quit - quit attach mode";
 
+	private final KubeClientFactory kubeClientFactory = DefaultKubeClientFactory.getInstance();
```
Comment (xintongsong) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/cli/KubernetesSessionCli.java:68]: minor: this can be a local variable

File: flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/TestingKubeClientFactory.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+
+/**
+ * Testing implementation of {@link KubeClientFactory}.
+ */
+public class TestingKubeClientFactory implements KubeClientFactory {
+
+	private TestingFlinkKubeClient.Builder flinkKubeClientBuilder;
```
Comment (xintongsong) [inline flink-kubernetes/src/test/java/org/apache/flink/kubernetes/kubeclient/TestingKubeClientFactory.java:33]: Should be `final`.

**Review by xintongsong — APPROVED** [review]
@SteNicholas 
No worries. Thanks for addressing my comments. The PR already looks quite good to me. 
I have only a few minor comments. I'll take over from here, and address my own comments while merging.

**Comment by xintongsong** [comment]
@SteNicholas 
I've updated the PR:
- Rebased onto latest `master` branch
- Your original changes are squashed into one main commit.
- My new changes are in the hotfix and fixup commits.

Please take a look at the changes. It there's no objections, I'll merge this once AZP gives green light.

**Comment by SteNicholas** [comment]
> @SteNicholas
> I've updated the PR:
> 
> * Rebased onto latest `master` branch
> * Your original changes are squashed into one main commit.
> * My new changes are in the hotfix and fixup commits.
> 
> Please take a look at the changes. It there's no objections, I'll merge this once AZP gives green light.

Yeah, I have taken a look at your changes. It's ok for me. Go ahead to merge this when AZP gives green light. Thanks for your detailed review and changes.
